src/cli/migrate/nodeVersion/index.ts

```diff
@@ -6,131 +6,122 @@ import fs from 'fs-extra';
 import { log } from '../../../utils/logging';
 import { createDestinationFileReader } from '../../configure/analysis/project';
 
+import {
+  isPatchableNodeVersion,
+  isPatchableServerlessVersion,
+  isPatchableSkubaType,
+} from './checks';
 import { getNodeTypesVersion } from './getNodeTypesVersion';
-import { validServerlessVersion, validSkubaType } from './packageJsonChecks';
 
-type SubPatch =
-  | (({ files: string; file?: never } | { file: string; files?: never }) & {
-      test?: RegExp;
-      replace: string;
-      id: string;
-    })
-  | Array<
-      ({ files: string; file?: never } | { file: string; files?: never }) & {
-        test?: RegExp;
-        replace: string;
-        id: string;
-      }
-    >;
+type FileSelector =
+  | { files: string; file?: never }
+  | { file: string; files?: never };
 
-const SHA_REGEX = /(?<=node.*)(@sha256:[a-f0-9]{64})/gm;
+type SubPatch = FileSelector & {
+  tests?: Array<() => Promise<boolean>>;
+  regex?: RegExp;
+  replace: string;
+};
 
-const subPatches: SubPatch[] = [
-  { id: 'nvmrc', file: '.nvmrc', replace: '<%- version %>\n' },
+const subPatches = ({
+  nodeVersion,
+  nodeTypesVersion,
+  ECMAScriptVersion,
+}: Versions): SubPatch[] => [
+  { file: '.nvmrc', replace: `${nodeVersion}\n` },
   {
-    id: 'Dockerfile-1',
     files: '**/Dockerfile*',
-    test: /^FROM(.*) (public.ecr.aws\/docker\/library\/)?node:[0-9.]+(@sha256:[a-f0-9]{64})?(\.[^- \n]+)?(-[^ \n]+)?( .+|)$/gm,
-    replace: 'FROM$1 $2node:<%- version %>$3$5$6',
+
+    regex:
+      /^FROM(.*) (public.ecr.aws\/docker\/library\/)?node:([0-9]+(?:\.[0-9]+(?:\.[0-9]+)?)?)(-[a-z0-9]+)?(@sha256:[a-f0-9]{64})?( .*)?$/gm,
+    replace: `FROM$1 $2node:${nodeVersion}$4$6`,
   },
   {
-    id: 'Dockerfile-2',
     files: '**/Dockerfile*',
-    test: /^FROM(.*) gcr.io\/distroless\/nodejs\d+-debian(.+)$/gm,
-    replace: 'FROM$1 gcr.io/distroless/nodejs<%- version %>-debian$2',
+    regex:
+      /^FROM(.*) gcr.io\/distroless\/nodejs\d+-debian(\d+)(@sha256:[a-f0-9]{64})?(\.[^- \n]+)?(-[^ \n]+)?( .+|)$/gm,
+    replace: `FROM$1 gcr.io/distroless/nodejs${nodeVersion}-debian$2$4$5$6`,
+  },
+
+  {
+    files: '**/serverless*.y*ml',
+    regex: /nodejs\d+.x/gm,
+    tests: [isPatchableServerlessVersion],
+    replace: `nodejs${nodeVersion}.x`,
   },
   {
-    id: 'serverless',
     files: '**/serverless*.y*ml',
-    test: /nodejs\d+.x/gm,
-    replace: 'nodejs<%- version %>.x',
+    regex: /node\d+/gm,
```

Comment:
For these very loose patterns, should we surround with word boundaries `\b` just as a bit of paranoia?